feat: no-cache attention in PyTorch workflow

qixiang-99 opened the pull request.

qixiang-99 committed:
init trtllm attn no cache

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
fix: fix the seq_len issue and attn metadata prepare for qwen reward model test

fix: fix minor bugs after rebase
Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: remove unnecessary debug logs and clean up commented code

refactor: update max_seq_len documentation and remove max_seq_len for decoder model contructor in PyTorchModelEngine
Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: update calculate_ref_result function to accept tensor inputs and mask type, enhance test_attention_no_cache to support FULL and CAUSAL masks

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: remove unused BERT attention metadata conversion method and add type assertion for no cache attention in PyTorchModelEngine

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: remove use_kv_cache parameter from attention function and related classes, update documentation for KV cache handling

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: implement setAttentionMaskType method for better mask type handling and remove unused conversion function

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: streamline KV cache handling by replacing direct member access with useKVCache method and simplify token per block assignment

remove Debug code.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
refactor: Resolve comments for Python code

Simplify no cache attention metadata preparation and streamline related attributes in TrtllmAttentionMetadata

Removed the private method for converting to no cache attention metadata and integrated its logic into the prepare method. Updated the test for BERT sequence classification to reflect these changes and ensure proper handling of attention metadata.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 commented:
/bot run

niukuo commented:
[PR_Github #498](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/498/) [ run ] triggered by Bot

niukuo commented:
[PR_Github #498](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/498/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #430](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/430/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run

qixiang-99 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #617](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/617/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #617](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/617/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #519](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/519/) completed with status: 'FAILURE'

symphonylyh reviewed (approved):
reviewed internally, approved.
Just need to add README and pass CI

qixiang-99 committed:
docs: Add is_dummy_attention field to attention metadata for simulation operations

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #622](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/622/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #622](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/622/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #524](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/524/) completed with status: 'FAILURE'

qixiang-99 committed:
refactor: add KVCacheParams to attention backend interface and import relevant metadata classes

Updated the attention backend interface to include KVCacheParams and imported TrtllmAttentionMetadata and VanillaAttentionMetadata in model_engine.py for enhanced functionality.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 committed:
fix: fix rebase format issue

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

qixiang-99 commented:
/bot run

qixiang-99 committed:
fix: extend attention mask type handling in MHARunnerFixedParams

Added support for additional attention mask types (BIDIRECTIONAL, BIDIRECTIONALGLM, BLOCKSPARSE) in the MHARunnerFixedParams structure to fix the mapping issue between ContextAttentionMaskType and AttentionMaskType

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

tensorrt-cicd commented:
[PR_Github #932](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/932/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #734](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/734/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #977](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/977/) [ run ] triggered by Bot

qixiang-99 commented:
/bot run

tensorrt-cicd commented:
[PR_Github #1036](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1036/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1036](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1036/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #800](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/800/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #1059](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1059/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1059](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1059/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #814 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/814/) completed with status: 'SUCCESS'

qixiang-99 commented:
/bot run --disable-fail-fast

qixiang-99 commented:
/bot run --stage-list "A30-CPP-1"

tensorrt-cicd commented:
[PR_Github #1137](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1137/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1137](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1137/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #856 `(Partly Tested)`](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/856/) completed with status: 'SUCCESS'

qixiang-99 committed:
fix: enhance attention mask type handling in TllmGenFmhaRunnerParams

Updated the setAttentionMaskType method to include a switch-case structure for better handling of attention mask types, ensuring proper mapping and error handling for invalid types.

Signed-off-by: Qixiang Lin <qixiangl@nvidia.com>

symphonylyh commented on cpp/tensorrt_llm/kernels/contextFusedMultiHeadAttention/fused_multihead_attention_common.h:
```diff
@@ -202,6 +202,7 @@ struct MHARunnerFixedParams
         case 2: // tensorrt_llm::kernels::AttentionMaskType::SLIDING_WINDOW_CAUSAL
             attentionMaskType = ContextAttentionMaskType::SLIDING_WINDOW_CAUSAL;
             break;
+        // NOTE: For BIDIRECTIONAL, BIDIRECTIONALGLM, BLOCKSPARSE context phase,CAUSAL mask is used
```
nit: missing a space before ,CAUSAL

qixiang-99 replied:
Good catch, [updated](https://github.com/NVIDIA/TensorRT-LLM/pull/3085/files#diff-60a870e4012f5122adb9a442138dc6ad3d7ea2b17817d6c8b506f90aaf5b533dR205), thanks!

symphonylyh commented on cpp/tensorrt_llm/kernels/trtllmGenKernels/fmha/fmhaRunnerParams.h:
```diff
@@ -238,8 +238,23 @@ struct TllmGenFmhaRunnerParams
-        mMaskType = static_cast<TrtllmGenAttentionMaskType>(maskType);
+        switch (maskType)
+        {
+            case 0:
```
do you want to also add the // kernel::AttentionMaskType::xxx here?

qixiang-99 replied:
Updated [here](https://github.com/NVIDIA/TensorRT-LLM/pull/3085/files#diff-f152f05f7879b557d289f85f4b2d06840ff20b855b5dd31adad0aa20e9114911R243-R254)

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #1146](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1146/) [ run ] triggered by Bot

qixiang-99 commented:
/bot kill

tensorrt-cicd commented:
[PR_Github #1146](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1146/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #863](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/863/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #1148](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1148/) [ kill ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1148](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1148/) [ kill ] completed with state `SUCCESS`
Successfully killed previous jobs for commit `e44b5f8`

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #1151](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1151/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1151](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1151/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #866](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/866/) completed with status: 'FAILURE'

tensorrt-cicd commented:
[PR_Github #1164](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1164/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1164](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1164/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #873](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/873/) completed with status: 'FAILURE'

qixiang-99 commented:
/bot run --disable-fail-fast

tensorrt-cicd commented:
[PR_Github #1167](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1167/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1167](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1167/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #875](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/875/) completed with status: 'SUCCESS'

symphonylyh commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #1182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1182/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #1182](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/1182/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #1167](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/1167/) for commit `c211717`
